drivers/console/aarch64/console.S

```diff
@@ -3,103 +3,240 @@
-	cbz	x0, init_fail
-	adrp	x3, console_base
-	str	x0, [x3, :lo12:console_base]
-	b	console_core_init
-init_fail:
-	ret
+	adrp	x14, console_list
+	str	xzr, [x14, :lo12:console_list]
+	b	console_core_register
 endfunc console_init
+#endif
+
+	/* -----------------------------------------------
+	 * console_info_t *console_unregister(
+	 *                        console_info_t *console)
+	 * Function to find a specific console in the list
+	 * of currently active consoles and remove it.
+	 * Should usually only be called by
+	 * declare_console expansions.
+	 * In: x0 - address of console structure to remove
+	 * Out: x0 - removed address, or NULL if not found
+	 * Clobber list: x0, x1, x14
+	 * -----------------------------------------------
+	 */
+func console_unregister
+	adrp	x14, console_list
+	add	x14, x14, :lo12:console_list	/* X14 = ptr to first struct */
+	ldr	x1, [x14]			/* X3 = first struct */
+
+unregister_loop:
+	cbz	x1, unregister_not_found
+	cmp	x0, x1
+	b.eq	unregister_found
+	ldr	x14, [x14]			/* X14 = next ptr of struct */
+	ldr	x1, [x14]			/* X1 = next struct */
+	b	unregister_loop
+
+unregister_found:
+	ldr	x1, [x1]			/* X1 = next struct */
+	str	x1, [x14]			/* prev->next = cur->next */
+	ret
+
+unregister_not_found:
+	mov	x0, #0				/* return NULL if not found */
+	ret
+endfunc console_unregister
 
 	/* -----------------------------------------------
 	 * void console_uninit(void)
-	 * Function to finish the use of console driver.
-	 * It sets the console_base as NULL so that any
-	 * further invocation of `console_putc` or
-	 * `console_getc` APIs would return error.
+	 * Function to switch from boot to runtime state.
+	 * The name is misleading for compatibility with
+	 * the legacy API -- this used to uninitialize all
+	 * consoles, but now it just quiesces consoles
+	 * that are not marked as "runtime".
+	 * Clobber list: x0, x1
 	 * -----------------------------------------------
 	 */
 func console_uninit
-	mov	x0, #0
-	adrp	x3, console_base
-	str	x0, [x3, :lo12:console_base]
+	mov	w0, #(1 << CONSOLE_FLAG_RUNTIME_SHIFT)
+	adrp	x1, console_state
+	strb	w0, [x1, :lo12:console_state]
 	ret
 endfunc console_uninit
 
+	/* -----------------------------------------------
+	 * int console_start_crash(void)
+	 * Function to switch to crash state. Returns 1 in
+	 * x0 for convenience with plat_crash_console_init
+	 * Out: x0 - always 1
+	 * Clobber list: x0, x1
+	 * -----------------------------------------------
+	 */
+func console_start_crash
+	mov	w0, #(1 << CONSOLE_FLAG_CRASH_SHIFT)
+	adrp	x1, console_state
+	strb	w0, [x1, :lo12:console_state]
+	mov	x0, #1
+	ret
+endfunc console_start_crash
+
 	/* ---------------------------------------------
-	 * int console_putc(int c)
-	 * Function to output a character over the
-	 * console. It returns the character printed on
-	 * success or -1 on error.
+	 * void console_putc(int c)
+	 * Function to output a character. Calls all
+	 * console's putc() handlers in succession.
```

Comment:
As I said in another comment, I'm not comfortable with removing the error codes, at least not as a side-effect of this patch. I'll try to canvas other opinions though. I agree this is source compatible but it's going to be important to keep function declarations and definitions the same, as this is a MISRA rule we want to comply with.